File: hazama/ui/taglist.py
import logging
from PySide.QtGui import *
from PySide.QtCore import *
from hazama.ui import font, refreshStyle
from hazama.ui.customobjects import DragScrollMixin
from hazama.ui.customwidgets import NElideLabel
from hazama.config import settings, db


class TagListDelegate(QStyledItemDelegate):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.h = font.default_m.height() + 8

    def paint(self, painter, option, index):
        x, y, w = option.rect.x(), option.rect.y(), option.rect.width()
        tag, count = index.data(Qt.DisplayRole), index.data(Qt.UserRole)
        if count is not None:
            count = str(count)
        painter.setFont(font.default)
        selected = bool(option.state & QStyle.State_Selected)
        textArea = QRect(x+4, y, w-8, self.h)
        if index.row() == 0:  # row 0 is always All(clear tag filter)
            painter.setPen(QColor(80, 80, 80))
            painter.drawText(textArea,
                             Qt.AlignVCenter | Qt.AlignLeft,
                             tag)
        else:
            painter.setPen(QColor(209, 109, 63))
            painter.drawLine(x, y, w, y)
            if selected:
                painter.setPen(QColor(181, 61, 0))
                painter.setBrush(QColor(250, 250, 250))
                painter.drawRect(x, y+1, w-1, self.h-2)
            # draw tag
            painter.setPen(QColor(20, 20, 20) if selected else
                           QColor(80, 80, 80))
            tag = font.default_m.elidedText(
                tag, Qt.ElideRight,
                w-12 if count is None else w-font.datetime_m.width(count)-12)
            painter.drawText(textArea, Qt.AlignVCenter | Qt.AlignLeft, tag)
            # draw tag count
            if count is not None:
                painter.setFont(font.datetime)
                painter.drawText(textArea, Qt.AlignVCenter | Qt.AlignRight, count)

    def createEditor(self, parent, option, index):
        # delegate will hold the reference to editor
        editor = QLineEdit(parent, objectName='tagListEdit')
        editor.oldText = index.data()
        return editor

    def updateEditorGeometry(self, editor, option, index):
        rect = option.rect
        rect.translate(1, 1)
        rect.setWidth(rect.width() - 2)
        rect.setHeight(rect.height() - 1)
        editor.setGeometry(rect)

    def sizeHint(self, option, index):
        return QSize(-1, self.h)


class TagListDelegateColorful(QItemDelegate):
    """ItemDelegate of theme 'colorful' for TagList. Using widget rendering."""
    class ItemWidget(QFrame):
        # almost the same as DiaryListDelegateColorful.ItemWidget
        def __init__(self, parent=None):
            super().__init__(parent, objectName='TagListItem')
            self._hLayout = QHBoxLayout(self)
            self._hLayout.setContentsMargins(0, 0, 0, 0)

            self.name = NElideLabel(self, objectName='TagListItemName')
            self.name.setAlignment(Qt.AlignRight)
            self.name.elideMode = Qt.ElideLeft
            self.count = QLabel(self, objectName='TagListItemCount')
            self.count.setSizePolicy(QSizePolicy.Maximum, QSizePolicy.Preferred)
            self._hLayout.addWidget(self.count)
            self._hLayout.addWidget(self.name)

        def setFixedWidth(self, w):
            if w != self.width():
                super().setFixedWidth(w)

    def __init__(self, parent=None):
        super().__init__(parent)
        self._itemW = self.ItemWidget()
        self._itemW.setFixedHeight(self._itemW.sizeHint().height())
        self._countEnabled = settings['Main'].getboolean('tagListCount')
        if not self._countEnabled: self._itemW.count.hide()

    def paint(self, painter, option, index):
        selected = bool(option.state & QStyle.State_Selected)
        active = bool(option.state & QStyle.State_Active)

        self._itemW.name.setText(index.data(Qt.DisplayRole))
        if self._countEnabled:
            countData = index.data(Qt.UserRole)
            self._itemW.count.setText(str(countData) if countData else '')
        self._itemW.setProperty('selected', selected)
        self._itemW.setProperty('active', active)
        refreshStyle(self._itemW)
        self._itemW.setFixedWidth(option.rect.width())

        painter.translate(option.rect.topLeft())
        self._itemW.render(painter, QPoint())
        painter.resetTransform()

    def sizeHint(self, option, index):
        return QSize(-1, self._itemW.height())

    def createEditor(self, parent, option, index):
        editor = QLineEdit(parent, objectName='tagListEdit')
        editor.setAlignment(self._itemW.name.alignment())
        editor.oldText = index.data()
        return editor

    def updateEditorGeometry(self, editor, option, index):
        editor.setGeometry(option.rect)


class TagList(DragScrollMixin, QListWidget):
    currentTagChanged = Signal(str)  # str is tag-name or ''
    tagNameModified = Signal(str)  # arg: newTagName

    def __init__(self, parent=None):
        DragScrollMixin.__init__(self)
        QListWidget.__init__(self, parent)
        self.setVerticalScrollMode(self.ScrollPerPixel)
        self.setupTheme()

        self.setUniformItemSizes(True)
        self.currentItemChanged.connect(self.onCurrentItemChanged)
        nextFunc = lambda: self.setCurrentRow(
            0 if self.currentRow() == self.count() - 1 else self.currentRow() + 1)
        preFunc = lambda: self.setCurrentRow((self.currentRow() or self.count()) - 1)
        self.nextSc = QShortcut(QKeySequence('Ctrl+Tab'), self, activated=nextFunc)
        self.preSc = QShortcut(QKeySequence('Ctrl+Shift+Tab'), self, activated=preFunc)

    def contextMenuEvent(self, event):
        # ignore "All" item. cursor must over the item
        index = self.indexAt(event.pos())
        if index.row() > 0:
            menu = QMenu()
            menu.addAction(QAction(self.tr('Rename'), menu,
                                   triggered=lambda: self.edit(index)))
            menu.exec_(event.globalPos())
            menu.deleteLater()

    def commitData(self, editor):
        newName = editor.text()
        if editor.isModified() and newName and ' ' not in newName:
            # editor.oldText is set in delegate
            db.change_tag_name(editor.oldText, newName)
            logging.info('tag [%s] changed to [%s]', editor.oldText, newName)
            super().commitData(editor)
            self.tagNameModified.emit(newName)

    def load(self):
        logging.debug('load Tag List')
        QListWidgetItem(self.tr('All'), self)
        self.setCurrentRow(0)
        itemFlag = Qt.ItemIsEditable | Qt.ItemIsSelectable | Qt.ItemIsEnabled
        if settings['Main'].getboolean('tagListCount'):
            for name, count in db.get_tags(count=True):
                item = QListWidgetItem(name, self)
                item.setFlags(itemFlag)
                item.setData(Qt.ToolTipRole, name)
                item.setData(Qt.UserRole, count)
        else:
            for name in db.get_tags(count=False):
                item = QListWidgetItem(name, self)
                item.setData(Qt.ToolTipRole, name)
                item.setFlags(itemFlag)

    def reload(self):
        if not self.isVisible():
            return

        try:
            currentTag = self.currentItem().data(Qt.DisplayRole)
        except AttributeError:  # no selection
            currentTag = None
        self.clear()
        self.load()
        if currentTag:
            try:
                item = self.findItems(currentTag, Qt.MatchFixedString)[0]
            except IndexError:
                item = self.item(0)
            self.setCurrentItem(item)

    def setupTheme(self):
        theme = settings['Main']['theme']
        d = {'colorful': TagListDelegateColorful}.get(theme, TagListDelegate)
        self.setItemDelegate(d())  # do not pass parent under PySide...
        # force items to be laid again
        self.setSpacing(self.spacing())

    def onCurrentItemChanged(self, currentItem):
        tag = currentItem.data(Qt.DisplayRole) if currentItem else ''
        # tag is '' if no selection
        self.currentTagChanged.emit('' if currentItem is self.item(0) else tag)

File: hazama/taglist.cpp
#include "taglist.h"
#include <QListWidgetItem>
#include <QShortcut>
#include <QMenu>
#include <QLineEdit>
#include <QContextMenuEvent>
#include "diarymodel.h"
#include "globals.h"
#include "db.h"
#include "taglistdelegates.h"

// tag list is simple, so don't use custom model
QVector<QPair<QString, int>> dbTags(bool getCount) {
    QVector<QPair<QString, int>> ret;
    QSqlQuery q;
    if (getCount)
        q.prepare("SELECT Tags.name, (SELECT COUNT(*) FROM Nikki_Tags "
                  "WHERE Nikki_Tags.tagid=Tags.id) AS count FROM Tags");
    else
        q.prepare("SELECT name FROM Tags");
    q.exec();
    while (q.next())
        ret.append(qMakePair(q.value(0).toString(), q.value(1).toInt()));
    return ret;
}


TagList::TagList(QWidget *parent) : QListWidget(parent) {
    nextSc = new QShortcut(QKeySequence("Ctrl+Tab"), this);
    connect(nextSc, &QShortcut::activated, [&](){ setCurrentRow((currentRow()+1) % count()); });
    preSc = new QShortcut(QKeySequence("Ctrl+Shift+Tab"), this);  // QKeySequence::PreviousChild has bug
    connect(preSc, &QShortcut::activated, [&](){ setCurrentRow((currentRow()-1+count()) % count()); });

    connect(this, &TagList::currentItemChanged, [&](QListWidgetItem *i) {
        // i become null after hiding the list
        emit currentTagChanged((i==nullptr || i==item(0)) ? "" : i->data(Qt::DisplayRole).toString());
    });
    setupDelegate();
}

void TagList::hideEvent(QHideEvent *) {
    clear();
}

void TagList::contextMenuEvent(QContextMenuEvent *e) {
    auto idx = currentIndex();
    if (idx.row() == 0)  // ignore "All" item. cursor must over the item
        return;
    QMenu menu;
    QAction rename(tr("Rename"), nullptr);
    connect(&rename, &QAction::triggered, [&](){ edit(idx); });
    menu.addAction(&rename);
    menu.exec(e->globalPos());
}

void TagList::commitData(QWidget *editor) {
    auto edit = qobject_cast<QLineEdit*>(editor);
    auto newName = edit->text();
    auto oldName = edit->property("oldText").toString();  // oldText is set in delegate's createEditor
    if (edit->isModified() && !newName.isEmpty() && !newName.contains(' ')) {
        DiaryModel::instance->changeTagName(oldName, newName);
        logI().quote() << "tag" << oldName << "changed to" << newName;
        QListWidget::commitData(editor);  // it will call delegate's setModelData
        emit tagNameModified(newName);
    }
}

void TagList::load() {
    Q_ASSERT(count() == 0);
    logD() << "load Tag List";
    new QListWidgetItem(tr("All"), this);
    setCurrentRow(0);

    bool getCount = settings->value("tagListCount").toBool();
    for (auto i : dbTags(getCount)) {
        auto item = new QListWidgetItem(i.first, this);
        item->setFlags(Qt::ItemIsEditable | Qt::ItemIsSelectable | Qt::ItemIsEnabled);
        item->setData(Qt::ToolTipRole, i.first);
        if (getCount)
            item->setData(Qt::UserRole, i.second);
    }
}

void TagList::reload() {
    if (!isVisible())
        return;
    auto currentTag = (currentItem() == nullptr) ? "" : currentItem()->data(Qt::DisplayRole).toString();
    clear();
    load();
    if (!currentTag.isEmpty()) {
        auto i = findItems(currentTag, Qt::MatchFixedString);
        setCurrentItem((i.isEmpty()) ? item(0) : i[0]);
    }
}

void TagList::setupDelegate() {
    auto theme = settings->value("theme").toString();
    if (theme == "colorful")
        setItemDelegate(new TagDelegateColorful(this));
    else
        setItemDelegate(new TagDelegate(this));
    if (isVisible())
        setSpacing(spacing());  // force items to be laid again
}
